File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> m;
        for(int val:nums)
            m[val]++;
        
        priority_queue<pair<int, int>> pq;
        for(auto pair: m)
            pq.push({pair.second, pair.first});
        
        vector<int> ans;
        int c = 0;
        while(!pq.empty()){
            if(++c>k) break;
            auto val = pq.top();
            pq.pop();
            ans.push_back(val.second);
        }
        return ans;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> freq = new HashMap<>();
        for (int num : nums)
            freq.put(num, freq.getOrDefault(num, 0) + 1);

        // min-heap by frequency
        PriorityQueue<Map.Entry<Integer,Integer>> heap = 
            new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());

        for (Map.Entry<Integer,Integer> entry : freq.entrySet()) {
            heap.offer(entry);
            if (heap.size() > k) heap.poll(); // keep only top k
        }

        int[] res = new int[k];
        for (int i = k - 1; i >= 0; i--) {
            res[i] = heap.poll().getKey();
        }
        return res;
    }
}